scripts/generate_setup.py

```diff
@@ -0,0 +1,192 @@
+
+import toml
+
+META = "METADATA.toml"
+
+# These constants may be adjusted, depending on convention we agree on.
+THIRD_PARTY_NAMESPACE = "stubs"
+PY2_NAMESPACE = "python2"
+SUFFIX = "-stubs"
+PY2_SUFFIX = "-python2-stubs"
+
+# After the typeshed refactoring PR lands, this will be not needed.
+OUTPUT_DIR = "out"
+
+SETUP_TEMPLATE = dedent("""
+from setuptools import setup
+
+name = "types-{distribution}"
+description = "Typing stubs for {distribution}"
+
+setup(name=name,
+      version="{version}",
+      description=description,
+      long_description=description,
+      url="https://github.com/python/typeshed",
+      install_requires={requires},
+      packages={packages},
+      package_data={package_data},
+      classifiers=[
+          "Typing :: Typed",
+      ]
+)
+""").lstrip()
+
+
+def find_stub_files(top: str) -> List[str]:
+    """Find all stub files for a given package, relative to package root.
+
+    Raise if we find any unknown file extensions during collection.
+    """
+    result = []
+    for root, _, files in os.walk(top):
+        for file in files:
+            if file.endswith(".pyi"):
+                result.append(os.path.relpath(os.path.join(root, file), top))
+            elif not file.endswith((".md", ".rst")):
+                # Allow including README docs, as some stubs have these (e.g. click).
+                raise ValueError("Only stub files are allowed")
+    return result
+
+
+def read_matadata(file: str) -> Dict[str, Any]:
+    """Parse metadata from file."""
+    with open(file) as f:
+        return dict(toml.loads(f.read()))
+
+
+def copy_stubs(distribution: str, dst: str, suffix: str) -> None:
+    """Copy stubs for given distribution to a temporary directory.
+
+    For packages change name by appending "-stubs" suffix (PEP 561),
+    also convert modules to trivial packages with a single __init__.pyi.
+    """
+    base_dir = os.path.join(THIRD_PARTY_NAMESPACE, distribution)
+    for entry in os.listdir(base_dir):
+        if os.path.isfile(os.path.join(base_dir, entry)):
+            if not entry.endswith(".pyi"):
+                continue
+            stub_dir = os.path.join(dst, entry.split(".")[0] + suffix)
+            os.mkdir(stub_dir)
+            shutil.copy(os.path.join(base_dir, entry), os.path.join(stub_dir, "__init__.pyi"))
+        else:
+            if entry == PY2_NAMESPACE:
+                # This is not really a package, but Python 2 stubs root.
+                # Packages there should be copied using a separate call to this
+                # function.
+                continue
+            else:
+                stub_dir = os.path.join(dst, entry + suffix)
+                shutil.copytree(os.path.join(base_dir, entry), stub_dir)
+
+        # We add original METADATA file in case some type-checking tool will want
+        # to use it. Note that since it is not easy to package it outside of a package,
+        # we copy it to every package in given distribution.
```

Comment:
METADATA.toml is needed to decide whether the stubs support Python 2/3, in particular.